### PR TITLE
599 - Extend Bookings

### PR DESCRIPTION
**Pull request by brownbearsoftware**
Add Extension tables to DB - extend the GET /premises/{id}/bookings endpoint to return the current extensions in the body.

Add POST /premises/{id}/bookings/{booking id}/extensions endpoint to create an extension:
 - Extension must be to a date later than the current departure date

### CODE REVIEW

**Comment by pezholio** [comment]
fedb60362ab615557fc4e3f030486520d6a3d461 seems to be an empty commit BTW 😃

File: src/main/resources/static/mini-manage-api-stubs.yml [hunk 1]
```diff
@@ -256,6 +256,59 @@ paths:
...
+          $ref: '#/components/responses/403Response'
+        404:
+          description: invalid premises ID or booking ID
+          content:
+            'application/json':
+              schema:
+                $ref: '#/components/schemas/Problem'
+        500:
+          $ref: '#/components/responses/500Response'
+      x-codegen-request-body-name: body
```
[the remark this answers is not in the document]
Comment (brownbearsoftware) [reply src/main/resources/static/mini-manage-api-stubs.yml:311]: Looks like it changes the generated parameters name: https://openapi-generator.tech/docs/swagger-codegen-migration/#body-parameter-name

Just added it for consistency as it was already on the others though

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/jpa/entity/BookingEntity.kt [hunk 2]
```diff
@@ -30,7 +30,7 @@ data class BookingEntity(
   val id: UUID,
   var crn: String,
   val arrivalDate: LocalDate,
-  val departureDate: LocalDate,
+  var departureDate: LocalDate,
```
Comment (pezholio) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/jpa/entity/BookingEntity.kt:33]: What is the significance of `val` vs `var`? Is it similar to `let` vs `const` in JS?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/controller/PremisesController.kt [hunk 3]
```diff
@@ -195,7 +198,35 @@ class PremisesController(
...
+        previousDepartureDate = booking.departureDate,
+        newDepartureDate = body.newDepartureDate,
+        notes = body.notes,
+        booking = booking
+      )
+    )
+
+    booking.departureDate = body.newDepartureDate
+    booking.extensions.add(extension)
+    bookingService.updateBooking(booking)
```
Comment (pezholio) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/controller/PremisesController.kt:227]: Should we do this in a transaction, so if one query fails we don't end up with a weird state where a booking has an extension, but that departure date hasn't been changed or vice versa?

**Review by pezholio — APPROVED** [review]
LGTM - just a coupla questions and a point about transaction safety

File: src/main/resources/static/mini-manage-api-stubs.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (pezholio) [reply src/main/resources/static/mini-manage-api-stubs.yml:311]: Cool 👍

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/jpa/entity/BookingEntity.kt (hunk 2 above)
Comment (brownbearsoftware) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/jpa/entity/BookingEntity.kt:33]: Yeah same concept as `let` vs `const` - `var` can be reassigned later whereas `val` cannot

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/controller/PremisesController.kt (hunk 3 above)
Comment (brownbearsoftware) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/controller/PremisesController.kt]: Good point, have ammended with a transaction

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/BookingService.kt [hunk 4]
```diff
@@ -8,16 +8,31 @@ import uk.gov.justice.digital.hmpps.approvedpremisesapi.jpa.entity.BookingEntity
...
+  private val cancellationRepository: CancellationRepository,
+  private val extensionRepository: ExtensionRepository
 ) {
   fun createBooking(bookingEntity: BookingEntity): BookingEntity = bookingRepository.save(bookingEntity)
+  fun updateBooking(bookingEntity: BookingEntity): BookingEntity = bookingRepository.save(bookingEntity)
   fun getBooking(id: UUID) = bookingRepository.findByIdOrNull(id)
   fun createArrival(arrivalEntity: ArrivalEntity): ArrivalEntity = arrivalRepository.save(arrivalEntity)
   fun createCancellation(cancellationEntity: CancellationEntity): CancellationEntity = cancellationRepository.save(cancellationEntity)
+
+  @Transactional
```
Comment (pezholio) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/BookingService.kt:29]: Oh wow! That's all you have to do? So it'll roll back if one or the other transaction fails?

Comment (brownbearsoftware) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/BookingService.kt:29]: Yeah - Spring creates a proxy subclass of BookingService when it detects this annotation then that's what actually gets injected into PremisesController - so when you call this method it actually calls the subclass first which adds the transaction handling/recovery which in turn calls this method

Comment (pezholio) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/BookingService.kt:29]: Nice!

**Review by pezholio — APPROVED** [review]

**Pull request merged**